packages/big-design/src/components/Modal/Modal.tsx

```diff
@@ -18,170 +18,146 @@ import {
-  private readonly headerUniqueId = uniqueId('modal_header_');
-
-  componentDidMount() {
-    const modalContainer = document.createElement('div');
-
-    document.body.appendChild(modalContainer);
-    this.setState({ modalContainer });
-  }
-
-  componentWillUnmount() {
-    const { initialBodyOverflowY, modalContainer } = this.state;
-
-    if (modalContainer) {
-      document.body.removeChild(modalContainer);
-    }
-
-    document.body.style.overflowY = initialBodyOverflowY;
-
-    this.returnFocus();
-  }
-
-  componentDidUpdate(prevProps: ModalProps) {
-    if (this.modalRef.current && !this.focusTrap) {
-      this.focusTrap = focusTrap(this.modalRef.current as HTMLElement, { initialFocus: this.modalRef.current });
-    }
-
-    // Check that the previous state was not open and is now open before auto focusing on modal
-    if (!prevProps.isOpen && this.props.isOpen) {
-      this.setState({ initialBodyOverflowY: document.body.style.overflowY || '' });
-    }
-
-    if (this.props.isOpen) {
-      document.body.style.overflowY = 'hidden';
-      this.focusTrap?.activate();
-    } else {
-      document.body.style.overflowY = this.state.initialBodyOverflowY;
-      this.focusTrap?.deactivate();
-      this.focusTrap = null;
-    }
-  }
-
-  render() {
-    const { backdrop, children, isOpen, variant } = this.props;
-    const { modalContainer } = this.state;
-
-    if (!(isOpen && modalContainer)) {
-      return null;
-    }
-
-    const modalContent = (
+export const Modal: React.FC<ModalProps> = typedMemo(
+  ({
+    actions,
+    backdrop = true,
+    children,
+    closeOnClickOutside = false,
+    closeOnEscKey = true,
+    header,
+    isOpen = false,
+    onClose = () => null,
+    variant = 'modal',
+  }) => {
+    const [internalTrap, setInternalTrap] = useState<FocusTrap | null>(null);
+    const [initialBodyOverflowY, setInitialBodyOverflowY] = useState('');
+    const [modalContainer, setModalContainer] = useState<HTMLDivElement | null>(null);
+    const headerUniqueId = useMemo(() => uniqueId('modal_header_'), []);
+    const modalRef = createRef<HTMLDivElement>();
+    const previousFocus = typeof document !== 'undefined' ? document.activeElement : null;
+
+    const returnFocus = () => {
+      const prevFocus = previousFocus as HTMLElement;
+
+      if (prevFocus && typeof prevFocus.focus === 'function') {
+        prevFocus.focus();
+      }
+    };
+
+    const onClickAway = (event: React.MouseEvent) => {
+      if (closeOnClickOutside && modalRef.current === event.target) {
+        onClose();
+      }
+    };
+
+    const onKeyDown = (event: React.KeyboardEvent) => {
+      if (closeOnEscKey && event.key === 'Escape') {
+        onClose();
+      }
+    };
+
+    useEffect(() => {
+      const container = document.createElement('div');
+
+      document.body.appendChild(container);
+      setModalContainer(container);
+    }, []);
+
+    useEffect(() => {
+      return () => {
```

Comment:
I'm not fully clear on why a function is returned from this useEffect() - would you mind explaining?

Reply:
Returning a function on a `useEffect` is roughly the same as `componentWillUnmount`. A common use-case for using this cleanup functionality is removing event listeners on the DOM:
```tsx
useEffect(() => {
  window.addEventListener('resize', handleResize);

  return () => {
    window.removeEventListener('resize', handleResize);
  };
});
```

This is extremely useful when components get added/removed from the DOM frequently.

Here's some more information on the effect cleanup: https://reactjs.org/docs/hooks-reference.html#cleaning-up-an-effect

---

In our use case we want the users focus to return back to their last focusable element. We use portals to add the modal onto the page so it'll mount/unmount every time it open/closes. Since we want to return focus, we run it on the effect cleanup.

Reply:
Oh thanks! I remember that now that you mention it. It's been a while since I've used the useEffect hook, will brush up on it again for my current work.